### PR TITLE
Requests/Events: Media Source/VLC

### LINKED ISSUE #494: OBS Media functions
##### Issue type
Feature request

##### Description
Since version 25 obs has new media functions, it would be very useful if we can use them.

##### Technical information
void obs_source_media_play_pause(obs_source_t *source, bool pause);
void obs_source_media_restart(obs_source_t *source);
void obs_source_media_stop(obs_source_t *source);
void obs_source_media_next(obs_source_t *source);
void obs_source_media_previous(obs_source_t *source);
int64_t obs_source_media_get_duration(obs_source_t *source);
int64_t obs_source_media_get_time(obs_source_t *source);
void obs_source_media_set_time(obs_source_t *source, int64_t ms);
enum obs_media_state obs_source_media_get_state(obs_source_t *source);
void obs_source_media_started(obs_source_t *source);
void obs_source_media_ended(obs_source_t *source);

### PR DESCRIPTION
**Pull request by tt2468**
Adds the following requests into a new category called `MediaControl`:
- `PlayPauseMedia`
- `RestartMedia`
- `StopMedia`
- `NextMedia`
- `PreviousMedia`
- `GetMediaDuration`
- `GetMediaTime`
- `SetMediaTime`
- `ScrubMedia`
- `GetMediaState`
- `GetMediaSourcesList`

Adds the following events into a new category called `Media`
- `MediaPlaying`
- `MediaPaused`
- `MediaRestarted`
- `MediaStopped`
- `MediaNext`
- `MediaPrevious`
- `MediaStarted`
- `MediaEnded`

Notes:
- I believe this requires at least OBS 25.0.8
- As with any large code change/addition, it would be nice to get some testers to bug test this. You should be able to download the plugin binaries through Azure pipelines.

Resolves #494

### CODE REVIEW

**Comment by tconley** [comment]
Would love to test this. i'm interested in creating a dock that gives playback controls and countown timer for media. idealy this would be for the program view as well as the preview view in studio mode.

**Comment by tt2468** [comment]
@tconley If you download the windows artifact from this page, it has an installer for obs-websocket with these features: https://dev.azure.com/Palakis/obs-websocket/_build/results?buildId=1243&view=artifacts&type=publishedArtifacts

**Comment by tconley** [comment]
Cool. I'm able to issue the requests like below.
`
{
"request-type":"PlayPauseMedia",
"sourceName":"movie1",
"playPause":false,
"message-id":"12324"
}
`

Should 'MediaPlaying' be return anytime that media is playing or only when i've triggered it via PlayPauseMedia?   For example if I click 'transition' in OBS Studio I get info back from websocket about that fact that SwitchScenes has happened, but I'm not getting back MediaPlaying

**Comment by tconley** [comment]
> Cool. I'm able to issue the requests like below.
> `{ "request-type":"PlayPauseMedia", "sourceName":"movie1", "playPause":false, "message-id":"12324" }`
> 
> Should 'MediaPlaying' be return anytime that media is playing or only when i've triggered it via PlayPauseMedia? For example if I click 'transition' in OBS Studio I get info back from websocket about that fact that SwitchScenes has happened, but I'm not getting back MediaPlaying

Figured that out, I am getting MediaStarted/MediaEnded events

**Comment by tt2468** [comment]
@tconley So the requests simple forward to internal requests in OBS, and the events simply forward signals coming from OBS.

In my testing, the media source had very different event behavior than the VLC source did. I honestly could not tell you whether something should or should not happen.

**Comment by tconley** [comment]
I was able to use these to come up with little webpage that finds all the media sources of typeID ffmpeg_source , added play/pause buttons and also a button to change the audio monitoring type for that source. super helpful to have this committed to obs-websocket!

**Comment by tt2468** [comment]
@tconley New request `GetMediaSourcesList` hot off the press.

**Comment by tconley** [comment]
@tt2468  `PlayRickRoleInCurrentScene` might be a good addition too

**Comment by PurpleMyst** [comment]
Hi! I need *exactly* this PR. Is it production ready? Can I just compile it myself and use it?

**Comment by tt2468** [comment]
@PurpleMyst I've just merged the main branch into the pr branch to update it to the latest features. Once the checks finish, you will be able to download the precompiled binaries for your appropriate OS here: https://dev.azure.com/Palakis/obs-websocket/_build/results?buildId=1282&view=results

It's in the "artifacts" section.

**Comment by PurpleMyst** [comment]
Ah thank you. I'm not familiar with the windows compilation story and was having a hard time figuring it out.

File: src/WSEvents.cpp [hunk 1]
```diff
@@ -1338,6 +1366,166 @@ void WSEvents::OnSourceFilterOrderChanged(void* param, calldata_t* data) {
 	self->broadcastUpdate("SourceFiltersReordered", fields);
 }
 
+/**
+ * A media source has started playing.
+ *
+ * Note: This event is only emitted when something actively controls the media/VLC source. In other words, the source will never emit this on its own naturally.
+ *
+ * @return {String} `sourceName` Source name
```
Comment (Palakis) [inline src/WSEvents.cpp:1374]: There's a `@return` line missing for `sourceTypeId`, here and in several other places where `getMediaSourceData` is used.

File: src/WSRequestHandler_MediaControl.cpp [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+		return request.failed("invalid request parameters");
+	}
+
+	OBSSourceAutoRelease source = obs_get_source_by_name(sourceName.toUtf8());
+	if (!source) {
+		return request.failed("specified source doesn't exist");
+	}
+
+	OBSDataAutoRelease response = obs_data_create();
+	obs_data_set_int(response, "timeStamp", obs_source_media_get_duration(source));
```
Comment (Palakis) [inline src/WSRequestHandler_MediaControl.cpp:221]: Doc says it is `mediaDuration` and not `timeStamp`. I prefer the former.

File: src/WSRequestHandler_MediaControl.cpp [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+	return request.success(response);
+}
+
+/**
+* Get the current timestamp of media in milliseconds. Supports ffmpeg and vlc media sources (as of OBS v25.0.8)
+*
+* @param {String} `sourceName` Source name.
+*
+* @return {int} `timeStamp` The time in milliseconds since the start of the media.
```
Comment (Palakis) [inline src/WSRequestHandler_MediaControl.cpp:230]: Timestamp is a single word, so no camelCase needed on it.

File: src/WSRequestHandler_MediaControl.cpp [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+	return request.success(response);
+}
+
+/**
+* List the media state of all media sources (vlc and media source)
+*
+* @return {Array<Object>} `mediaSources` Array of sources
+* @return {String} `mediaSources.*.sourceName` Unique source name
+* @return {String} `mediaSources.*.sourceTypeId` Non-unique source internal type (a.k.a `ffmpeg_source` or `vlc_source`)
```
Comment (Palakis) [inline src/WSRequestHandler_MediaControl.cpp:363]: Type id is called "source kind" everywhere else so I'd rather stay consistent with that.
```suggestion
* @return {String} `mediaSources.*.sourceKind` Unique source internal type (a.k.a `ffmpeg_source` or `vlc_source`)
```

File: src/WSRequestHandler_MediaControl.cpp [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+	OBSDataArrayAutoRelease sourcesArray = obs_data_array_create();
+
+	auto sourceEnumProc = [](void* privateData, obs_source_t* source) -> bool {
+		obs_data_array_t* sourcesArray = (obs_data_array_t*)privateData;
+
+		QString sourceTypeId = obs_source_get_id(source);
+		if (isMediaSource(sourceTypeId)) {
+			OBSDataAutoRelease sourceData = obs_data_create();
+			obs_data_set_string(sourceData, "sourceName", obs_source_get_name(source));
+			obs_data_set_string(sourceData, "sourceTypeId", sourceTypeId.toUtf8());
```
Comment (Palakis) [inline src/WSRequestHandler_MediaControl.cpp:382]:
```suggestion
			obs_data_set_string(sourceData, "sourceKind", sourceTypeId.toUtf8());
```

**Pull request merged**